lib/sharing/ConsumeSharedPlugin.js

```diff
@@ -252,6 +256,62 @@ class ConsumeSharedPlugin {
 								requiredVersion
 							}
 						);
+						// Check if exclusion criteria is provided
+						if (config.exclusionCriteria && config.exclusionCriteria.version) {
+							// Without a resolved import, we cannot proceed with exclusion check
+							if (!importResolved) {
+								return consumedModule;
+							}
+							// If a fallback version is provided by the user, use that to check
+							if (config.exclusionCriteria.fallbackVersion) {
+								if (
+									satisfy(
+										parseRange(config.exclusionCriteria.version),
+										config.exclusionCriteria.fallbackVersion
+									)
+								) {
+									return undefined;
+								}
+								return consumedModule;
+							}
+							return new Promise(resolve => {
+								// Get the closest package.json file corresponding to the resolved import.
+								// This version will be used for checking whether the dependency must be
+								// excluded from consumption (in a share context). If such a version does
+								// not exist in the package.json file, then we use the "version" property
+								// set by users in the share config, which is passed here as fallbackVersion
+								// a.k.a the version used to register the package in the scope.
+								getDescriptionFile(
+									compilation.inputFileSystem,
+									path.dirname(importResolved),
+									["package.json"],
+									(err, result) => {
+										if (err) {
+											return resolve(consumedModule);
+										}
+										const { data } = result;
+										if (
+											!data ||
+											(data && !data.version) ||
+											(data && data.name !== request)
+										) {
```

Comment:
Check that the `package.json` name matches the requested module. Otherwise it can resolve to the root `package.json` of the project.